src/dataset.rs

```diff
@@ -508,6 +511,84 @@ impl Dataset {
         }
         Ok(Transaction::new(self))
     }
+
+    /// Execute a SQL query against the Dataset. It is equivalent to calling
+    /// [`GDALDatasetExecuteSQL`](https://gdal.org/api/raster_c_api.html#_CPPv421GDALDatasetExecuteSQL12GDALDatasetHPKc12OGRGeometryHPKc).
+    /// Returns a `ResultSet`, which can be treated just as any other [`Layer`], or `None` if the
+    /// query does not have a `ResultSet`, such as `ALTER TABLE`, etc. Note that this is different
+    /// than a query which has an *empty* `ResultSet`.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # use gdal::Dataset;
+    /// # use std::path::Path;
+    /// #
+    /// let ds = Dataset::open(Path::new("fixtures/roads.geojson")).unwrap();
+    /// let query = "SELECT kind, is_bridge, highway FROM roads WHERE highway = 'pedestrian'";
+    /// let result_set = ds.execute_sql(query, None, None).unwrap().unwrap();
+    ///
+    /// assert_eq!(10, result_set.feature_count());
+    ///
+    /// for feature in result_set.features() {
+    ///     let highway = feature
+    ///         .field("highway")
+    ///         .unwrap()
+    ///         .unwrap()
+    ///         .into_string()
+    ///         .unwrap();
+    ///
+    ///     assert_eq!("pedestrian", highway);
+    /// }
+    /// ```
+    pub fn execute_sql<S: AsRef<str>>(
+        &self,
+        query: S,
+        spatial_filter: Option<&Geometry>,
+        dialect: Option<S>,
+    ) -> Result<Option<ResultSet>> {
+        let query = CString::new(query.as_ref())?;
+        let mut filter_geom: OGRGeometryH = std::ptr::null_mut();
+
+        let dialect_c_str = match dialect {
+            None => None,
+            Some(d) => Some(CString::new(d.as_ref())?),
+        };
+
+        let dialect_ptr = match dialect_c_str {
+            None => std::ptr::null(),
+            Some(d) => d.as_ptr(),
+        };
+
+        if let Some(spatial_filter) = spatial_filter {
+            filter_geom = unsafe { spatial_filter.c_geometry() };
+        }
+
+        let c_dataset = unsafe { self.c_dataset() };
+
+        unsafe { gdal_sys::CPLErrorReset() };
```

Comment:
Is this really needed? From the docs I gather that one needs it call it after handling an error, presumably in the GDAL code.

Reply:
This is just defensive, in case for some reason by the time this function is called, some previous error had not been cleared. If this was the case, it would return an error even when the query was successful, and the error would be completely unrelated to the query, which would be very confusing.

Reply:
What I meant is that GDAL itself resets the last error when you call into it through a public API. `GDALOpenEx` does that, the GeoJSON reader does it when needed and so on. You could argue that GDAL itself might forget to call it, but it doesn't seem worth it littering the code with these calls.

Reply:
Just taking a poke through [`gdaldataset.cpp`](https://github.com/OSGeo/gdal/blob/master/gdal/gcore/gdaldataset.cpp), it appears that `CPLErrorReset` is called in `GDALOpenEx`, and also `GDALDataset::CopyLayer`, but not in `GDALExecuteSQL`. It's possible it is called somewhere further down in the stack.

However, you probably have more experience working with the GDAL code base than me, so if you're not worried, I'll defer to your judgment and remove it.

Reply:
It's actually called -- in your SQLite test -- when SQLite opens the dataset, as it goes through `GDALOpenEx` (I checked using GDB).

I don't really have experience working on GDAL itself but it would make sense for GDAL to clear the last error. On the other hand, the Python bindings seem to call `CPLErrorReset`, so maybe you're right and we should follow suit.

Reply:
Right, it probably calls it when the data set is opened, but does it call it after it enters the `ExecuteSQL` function? It seems the case where it could cause problems is if the user opens a Dataset, calls one or more other functions, and the last one errors, but the user doesn't check, and then calls `ExecuteSQL`.

Now whether this would happen through the use of this specific crate might be a different matter, since it looks like everywhere that errors calls `_last_cpl_err`, which does reset the error state before returning. Although this does assume that `_last_cpl_err` is called at every site where an error could possibly happen.

I'll leave it up to your guys' judgment.

Reply:
> Right, it probably calls it when the data set is opened, but does it call it after it enters the ExecuteSQL function?

Yes, because it opens the dataset again:

```
#0  0x00007ffff7a749a0 in CPLErrorReset () from /usr/lib/libgdal.so.26
#1  0x00007ffff7582f62 in GDALOpenEx () from /usr/lib/libgdal.so.26
#2  0x00007ffff797a427 in OGROpen () from /usr/lib/libgdal.so.26
#3  0x00007ffff79ae9ef in ?? () from /usr/lib/libgdal.so.26
#4  0x00007ffff4b9ebd2 in sqlite3VdbeExec () from /usr/lib/libsqlite3.so.0
#5  0x00007ffff4ba50b8 in sqlite3_step () from /usr/lib/libsqlite3.so.0
#6  0x00007ffff79a866d in OGRSQLiteTableLayer::EstablishFeatureDefn(char const*) () from /usr/lib/libgdal.so.26
#7  0x00007ffff79a9ca7 in OGRSQLiteTableLayer::Initialize(char const*, int, int) () from /usr/lib/libgdal.so.26
#8  0x00007ffff798983e in OGRSQLiteDataSource::OpenTable(char const*, int) () from /usr/lib/libgdal.so.26
#9  0x00007ffff7989966 in OGRSQLiteDataSource::GetLayerByName(char const*) () from /usr/lib/libgdal.so.26
#10 0x00007ffff79899ea in OGRSQLiteDataSource::GetLayerWithGetSpatialWhereByName(char const*) () from /usr/lib/libgdal.so.26
#11 0x00007ffff799d31b in OGRSQLiteSelectLayerCommonBehaviour::GetBaseLayer(unsigned long&) () from /usr/lib/libgdal.so.26
#12 0x00007ffff799d539 in OGRSQLiteSelectLayerCommonBehaviour::BuildSQL() () from /usr/lib/libgdal.so.26
#13 0x00007ffff799e7e9 in OGRSQLiteSelectLayerCommonBehaviour::SetSpatialFilter(int, OGRGeometry*) () from /usr/lib/libgdal.so.26
#14 0x00007ffff79953e3 in OGRSQLiteExecuteSQL(GDALDataset*, char const*, OGRGeometry*, char const*) () from /usr/lib/libgdal.so.26
```

> Now whether this would happen through the use of this specific crate might be a different matter, since it looks like everywhere that errors calls _last_cpl_err, which does reset the error state before returning. Although this does assume that _last_cpl_err is called at every site where an error could possibly happen.

Yeah, the user could still call `gdal-sys` bindings without resetting the last error.

> I'll leave it up to your guys' judgment.

Sorry for the ping-pong with this, but if the Python bindings reset the last error, I think it's safe to do the same for now.

Reply:
Ah, right, it would create the virtual table, and then open that as a Dataset. So then the case where this is possible would be non-SQLite dialects.

Okay, I've added it back. I think that takes care of all the review comments.